### PR TITLE
[Feature/ChangedFiles]#125 Add support for getting changedFiles from PR as resource

### PR DESCRIPTION
**Pull request by holger-wg2**
Queries the V4 Github API for a list of ChangedFiles in a PR and writes them alongside with version.json and metadata.json as a plain text file called changedFiles into the .git/resource folder

### CODE REVIEW

File: github.go [hunk 1]
```diff
@@ -216,6 +216,13 @@ func (m *GithubClient) GetPullRequest(prNumber, commitRef string) (*PullRequest,
 						}
 					}
 				} `graphql:"commits(last:$commitsLast)"`
+				Files struct {
+					Edges []struct {
+						Node struct {
+							FilesChangedObject
+						}
+					}
+				} `graphql:"files(last:$commitsLast)"`
```
[the remark this answers is not in the document]
Comment (holger-wg2) [reply github.go]: According to the documentation it returns the last n number of files from the file list in the PR

File: in.go [hunk 2]
```diff
@@ -93,6 +93,19 @@ func Get(request GetRequest, github Github, git Git, outputDir string) (*GetResp
...
 
+	if len(pull.Files) != 0 {
+		var fl []byte
+
+		for _, v := range pull.Files {
+			fl = append(fl, []byte(v.Path+"\n")...)
+		}
+
+		// Create List with changed files
+		if err := ioutil.WriteFile(filepath.Join(path, "changedFiles"), fl, 0644); err != nil {
```
[the remark this answers is not in the document]
Comment (holger-wg2) [reply in.go]: Done

File: models.go [hunk 3]
```diff
@@ -73,7 +73,8 @@ func NewVersion(p *PullRequest) Version {
 // PullRequest represents a pull request and includes the tip (commit).
 type PullRequest struct {
 	PullRequestObject
-	Tip CommitObject
+	Tip   CommitObject
+	Files []FilesChangedObject
```
[the remark this answers is not in the document]
Comment (holger-wg2) [reply models.go]: Done

File: models.go [hunk 4]
```diff
@@ -104,3 +105,9 @@ type CommitObject struct {
 		}
 	}
 }
+
+// FilesChangedObject represents the GraphQL FilesChanged node.
+// https://developer.github.com/v4/object/pullrequestchangedfile/
+type FilesChangedObject struct {
```
[the remark this answers is not in the document]
Comment (holger-wg2) [reply models.go]: Done

File: github.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (holger-wg2) [reply github.go]: Added a new changedFilesLast variable making it configurable

**Comment by holger-wg2** [comment]
> Thanks for the PR, this looks great! Aside from the nitpicks in my comments, I'm wondering how `last` works for listing PR files in Github V4, and thinking that we'll probably be required to do pagination in order to have a complete list of modified files and not just some of them. If we do pagination however, we'll also have to make this functionality optional via a flag, since it would increase the costs of the operation (ratelimits).
> 
> If we make this optional, it might make sense to just move the implementation to [ListModifiedFiles()](https://github.com/telia-oss/github-pr-resource/blob/master/github.go#L153) so we can use it for `path` and `ignore_paths` in `check` as well?

I do agree with you, we would need pagination. Right now it is limited to about the 100 last files in the list of changed Files.
However the number of commits then would also need to be paginated, which is not done right now.

**Comment by itsdalmo** [comment]
> However the number of commits then would also need to be paginated, which is not done right now.

Does this mean that the list of modified files is per commit, and not for the PR as a whole? I.e., to get a complete list for the entire PR we would have to paginate all the modified files for every commit?

**Comment by holger-wg2** [comment]
Sorry for being unclear. The List of changed files is per PR.

I just wanted to point out the big TODO in the code about pagination of commits.

**Comment by itsdalmo** [comment]
> Sorry for being unclear. The List of changed files is per PR.
> I just wanted to point out the big TODO in the code about pagination of commits.

Ah I see, thanks for clarifying. The TODO about pagination would be nice to fix also, but I think for now we can just focus on having pagination when listing the modified files since it is more likely that users will have 100+ modified files? What do you think about making the functionality optional though, and maybe moving the implementation to `ListModifiedFiles()`?

**Comment by holger-wg2** [comment]
I will look into pagination. However I would rather not move it into ListModifiedFiles() as the information is really most useful when the repo gets checked out and the information is written to disk. I think check stay a relatively short operation.

**Comment by holger-wg2** [comment]
Rebased on current master

**Comment by holger-wg2** [comment]
Quite a big change, but the fetching of files now supports pagination

**Comment by holger-wg2** [comment]
@itsdalmo Getting the changed file list from the PR is now optional and supports pagination.

File: github.go [hunk 5]
```diff
@@ -126,7 +127,7 @@ func (m *GithubClient) ListOpenPullRequests() ([]*PullRequest, error) {
 		"prFirst":         githubv4.Int(100),
 		"prStates":        []githubv4.PullRequestState{githubv4.PullRequestStateOpen},
 		"prCursor":        (*githubv4.String)(nil),
-		"commitsLast":     githubv4.Int(1),
+		"commitsLast":     githubv4.Int(100),
```
Comment (itsdalmo) [inline github.go]: This is intended to be just the very last commit (i.e. 1 and not 100) since the function is just used for `check`.

File: github.go [hunk 6]
```diff
@@ -198,8 +199,66 @@ func (m *GithubClient) PostComment(prNumber, comment string) error {
...
+		Repository struct {
+			PullRequest struct {
+				Files struct {
+					Edges []struct {
+						Node struct {
+							ChangedFileObject
+						}
+					} `graphql:"edges"`
+					PageInfo struct {
+						PageInfoObject
```
Comment (itsdalmo) [inline github.go]: Instead of a separate `PageInfoObject` added to `models.go`, you could just use an inline struct as done in other cases where pagination is implemented. I.e., just inline the struct:

```go
PageInfo struct {
    EndCursor   githubv4.String
    HasNextPage bool
}
```

File: github.go [hunk 7]
```diff
@@ -198,8 +199,66 @@ func (m *GithubClient) PostComment(prNumber, comment string) error {
...
+					} `graphql:"pageInfo"`
+				} `graphql:"files(first:$changedFilesFirst, after: $changedFilesEndCursor)"`
+			} `graphql:"pullRequest(number:$prNumber)"`
+		} `graphql:"repository(owner:$repositoryOwner,name:$repositoryName)"`
+	}
+
+	offset := ""
+	fetch := true
+
+	for fetch {
```
Comment (itsdalmo) [inline github.go]: You don't need to use a `while true` type loop here, it's enough to just do `for { }` and `break` when there are there is no more pagination to be done. Please use a loop like in `ListOpenPullRequests()`; might be a difference of opinion, but I'd prefer if we just stick with one type of loop if possible 😅

File: github.go [hunk 8]
```diff
@@ -198,8 +199,66 @@ func (m *GithubClient) PostComment(prNumber, comment string) error {
...
+		}
+
+	}
+
+	return cfo, nil
+}
+
 // GetPullRequest ...
-func (m *GithubClient) GetPullRequest(prNumber, commitRef string) (*PullRequest, error) {
+func (m *GithubClient) GetPullRequest(prNumber, commitRef string, changedFileQuery bool) (*PullRequest, error) {
```
Comment (itsdalmo) [inline github.go]: I think it makes sense to implement this in `in.go` instead of adding the condition to this function signature, since it ends up being two API calls anyway.

File: in.go [hunk 9]
```diff
@@ -109,9 +123,10 @@ func Get(request GetRequest, github Github, git Git, outputDir string) (*GetResp
 
 // GetParameters ...
 type GetParameters struct {
-	SkipDownload    bool   `json:"skip_download"`
-	IntegrationTool string `json:"integration_tool"`
-	GitDepth        int    `json:"git_depth"`
+	SkipDownload      bool   `json:"skip_download"`
+	IntegrationTool   string `json:"integration_tool"`
+	GitDepth          int    `json:"git_depth"`
+	ChangedFilesQuery bool   `json:"generate_changed_file_list"`
```
Comment (itsdalmo) [inline in.go]: `ChangedFilesQuery` should be `GenerateChangedFileList` to match whatever parameter name we go for (makes it easier to read the code).

File: in.go [hunk 10]
```diff
@@ -99,6 +99,20 @@ func Get(request GetRequest, github Github, git Git, outputDir string) (*GetResp
 		if err := ioutil.WriteFile(filepath.Join(path, filename), content, 0644); err != nil {
 			return nil, fmt.Errorf("failed to write metadata file %s: %s", filename, err)
 		}
+
+	}
+
+	if len(pull.ChangedFiles) != 0 {
```
Comment (itsdalmo) [inline in.go]: Ref my comment about the function signature of `GetPullRequest`. I think it would make sense if this if statement checked `if request.Params.ChangedFilesQuery` (to be renamed) and listed the changed files right in the if statement. Doesn't even need to add it to the PR object since it's not used anywhere else yet. Thoughts?

File: github.go (hunk 5 above)
Comment (holger-wg2) [reply github.go]: Yes of course. Sorry was confused here because of getting 100 changed files in one go. Will fix. Thanks for catching that.

File: github.go (hunk 6 above)
Comment (holger-wg2) [reply github.go]: I was thinking about reusability for adding pagination on commits at one point if needed

File: README.md [hunk 11]
```diff
@@ -60,11 +60,12 @@ generate notifications over the webhook. So if you have a repository with little
...
-| `integration_tool` | No       | `rebase` | The integration tool to use, `merge`, `rebase` or `checkout`. Defaults to `merge`. |
-| `git_depth`        | No       | `1`      | Shallow clone the repository using the `--depth` Git option                        |
+| Parameter                     | Required | Example  | Description                                                                        |
+|-------------------------------|----------|----------|------------------------------------------------------------------------------------|
+| `skip_download`               | No       | `true`   | Use with `get_params` in a `put` step to do nothing on the implicit get.           |
+| `integration_tool`            | No       | `rebase` | The integration tool to use, `merge`, `rebase` or `checkout`. Defaults to `merge`. |
+| `git_depth`                   | No       | `1`      | Shallow clone the repository using the `--depth` Git option                        |
+| `generate_changed_file_list`  | No       | `true`   | Generate a list of changed files and save alongside metadata                       |
```
[the remark this answers is not in the document]
Comment (holger-wg2) [reply README.md]: yes, makes sense

File: github.go (hunk 7 above)
Comment (holger-wg2) [reply github.go]: Sure

File: in.go (hunk 9 above)
Comment (holger-wg2) [reply in.go]: Sure

File: github.go (hunk 6 above)
Comment (holger-wg2) [reply github.go]: Will do what is used in other places for pagination

File: github.go (hunk 8 above)
Comment (holger-wg2) [reply github.go]: Sure

File: in.go (hunk 10 above)
Comment (holger-wg2) [reply in.go]: I moved it around. This actually also made the test code slightly smaller, however conceptually it would have fit into the PR object quite well

**Comment by holger-wg2** [comment]
@itsdalmo I think I addressed now all your comments

File: github.go (hunk 5 above)
Comment (itsdalmo) [inline github.go]: This should be `1` 😅

File: github.go [hunk 12]
```diff
@@ -224,20 +281,23 @@ func (m *GithubClient) GetPullRequest(prNumber, commitRef string) (*PullRequest,
 		"repositoryOwner": githubv4.String(m.Owner),
 		"repositoryName":  githubv4.String(m.Repository),
 		"prNumber":        githubv4.Int(pr),
-		"commitsLast":     githubv4.Int(100),
+		"commitsLast":     githubv4.Int(1),
```
Comment (itsdalmo) [inline github.go]: This one should be `100` 😂

File: github.go [hunk 13]
```diff
@@ -224,20 +281,23 @@ func (m *GithubClient) GetPullRequest(prNumber, commitRef string) (*PullRequest,
...
 	}
 
 	// TODO: Pagination - in case someone pushes > 100 commits before the build has time to start :p
 	if err := m.V4.Query(context.TODO(), &query, vars); err != nil {
 		return nil, err
 	}
+
 	for _, c := range query.Repository.PullRequest.Commits.Edges {
+		prObj := PullRequest{
```
Comment (itsdalmo) [inline github.go]: Please revert the changes in this `for` loop this so there is no diff, since there are no real code changes anymore after you've moved your list function.

File: in.go [hunk 14]
```diff
@@ -109,9 +128,10 @@ func Get(request GetRequest, github Github, git Git, outputDir string) (*GetResp
 
 // GetParameters ...
 type GetParameters struct {
-	SkipDownload    bool   `json:"skip_download"`
-	IntegrationTool string `json:"integration_tool"`
-	GitDepth        int    `json:"git_depth"`
+	SkipDownload            bool   `json:"skip_download"`
+	IntegrationTool         string `json:"integration_tool"`
+	GitDepth                int    `json:"git_depth"`
+	GenerateChangedFileList bool   `json:"list_changed_files"`
```
Comment (itsdalmo) [inline in.go]: The struct field should mirror the JSON key name, so please rename to `ListChangedFiles`

File: in_test.go [hunk 15]
```diff
@@ -266,7 +349,7 @@ func createTestPR(count int, baseName string, skipCI bool, isCrossRepo bool) *re
 		m = "[skip ci]" + m
 	}
 
-	return &resource.PullRequest{
+	retpr := resource.PullRequest{
```
Comment (itsdalmo) [inline in_test.go]: Please revert this change as well so there is no diff, since there are no longer any code changes (and no reason to declare `reptr` instead of just returning the PR object directly.

File: models.go [hunk 16]
```diff
@@ -73,7 +73,7 @@ func NewVersion(p *PullRequest) Version {
 // PullRequest represents a pull request and includes the tip (commit).
 type PullRequest struct {
 	PullRequestObject
-	Tip CommitObject
+	Tip          CommitObject
```
Comment (itsdalmo) [inline models.go]: This looks like it might be missing a `go fmt`?

**Comment by holger-wg2** [comment]
The travis build timed out out of unknown reasons. Please retrigger @itsdalmo

File: github.go [hunk 17]
```diff
@@ -198,6 +199,62 @@ func (m *GithubClient) PostComment(prNumber, comment string) error {
...
+			"changedFilesFirst":     githubv4.Int(100),
+			"changedFilesEndCursor": githubv4.String(offset),
+		}
+
+		if err := m.V4.Query(context.TODO(), &filequery, vars); err != nil {
+			return nil, err
+		}
+
+		for _, f := range filequery.Repository.PullRequest.Files.Edges {
+			cfo = append(cfo, ChangedFileObject{Path: f.Node.Path})
```
Comment (itsdalmo) [inline github.go]: I think you might be able to just append the `ChangedFileObject` directly here instead of redeclaring it? I.e.:

```go
cfo = append(cfo, f.Node.ChangedFileObject)
```

Or am I missing something?

File: github.go (hunk 5 above)
Comment (holger-wg2) [reply github.go]: :-( yeah ... will fix

File: github.go (hunk 12 above)
Comment (holger-wg2) [reply github.go]: :-( yeah ... will fix

File: github.go (hunk 13 above)
Comment (holger-wg2) [reply github.go]: done

File: in.go (hunk 14 above)
Comment (holger-wg2) [reply in.go]: done

File: in_test.go (hunk 15 above)
Comment (holger-wg2) [reply in_test.go]: done

File: models.go (hunk 16 above)
Comment (holger-wg2) [reply models.go]: done

File: github.go (hunk 17 above)
Comment (holger-wg2) [reply github.go]: I think because append needs to resize the slice this is actually needed

**Comment by holger-wg2** [comment]
Fixed all except the append stuff. Good enough? @itsdalmo
